File: code/c++/655. 输出二叉树.cpp
// 655. Print Binary Tree
class Solution {
public:
  vector<vector<string>> printTree(TreeNode *root) {
    int rows = get_height(root);
    int cols = pow(2, m) - 1; // cols is all nodes of tree
    cout << "rows:" << rows << " cols:" << cols << endl;
    vector<vector<string>> res(rows, vector<string>(cols, ""));
    dfs(root, res, 0, 0, n - 1);
    return res;
  }

  void dfs(TreeNode *root, vector<vector<string>> &res, int row, int start,
           int end) {
    if (!root || (start > end))
      return;
    int mid = (start + end) >> 1;
    res[row][mid] = to_string(root->val);
    dfs(root->left, res, row + 1, start, mid);
    dfs(root->right, res, row + 1, mid + 1, end);
  }

  int get_height(TreeNode *root) {
    if (!root)
      return 0;
    return std::max(get_height(root->left), get_height(root->right)) + 1;
  }
};

File: code/go/655. 输出二叉树.go
// 655. 输出二叉树
package main

import (
	"math"
	"strconv"
)

/** 655. Print Binary Tree
Input:
     1
    / \
   2   3
    \
     4
Output:
[["", "", "", "1", "", "", ""],
 ["", "2", "", "", "", "3", ""],
 ["", "", "4", "", "", "", ""]]

*/
func printTree(root *TreeNode) [][]string {

	rows := getHeight(root)
	cols := math.Pow(2, rows) - 1 // all nodes of tree

	var res [][]string //nil point
	for i := 0; i < int(rows); i++ {
		res = append(res, make([]string, int(cols)))
	}

	dfsPrintTree(root, 0, 0, int(cols)-1, res)

	return res

}

func dfsPrintTree(root *TreeNode, level int, start int, end int, res [][]string) {
	if root == nil || start > end {
		return
	}

	//print root
	mid := (start + end) / 2
	res[level][mid] = strconv.Itoa(root.Val)

	//print left subTree
	dfsPrintTree(root.Left, level+1, start, mid-1, res)
	//print right subTree
	dfsPrintTree(root.Right, level+1, mid+1, end, res)

}

//height of tree
func getHeight(root *TreeNode) float64 {
	if root == nil {
		return 0
	}
	left := getHeight(root.Left)
	rgiht := getHeight(root.Right)
	return math.Max(left, rgiht) + 1
}

/**
func main() {
	tree := CreateTestData("[1,2,3,4]")
	Print(tree)
	fmt.Println("Hello World!")
}**/
